src/deploy.sh

```diff
@@ -23,169 +23,336 @@ include "$KW_LIB_DIR/signal_manager.sh"
+    statistics_manager "uninstall" "$runtime"
+    return "$?"
+  fi
 
   if ! is_kernel_root "$PWD"; then
     complain "Execute this command in a kernel tree."
     exit 125 # ECANCELED
   fi
 
-  flag=${flag:-""}
-
-  case "$target" in
-    1) # VM_TARGET
-      distro=$(detect_distro "${configurations[mount_point]}/")
-
-      if [[ "$distro" =~ "none" ]]; then
-        complain "Unfortunately, there's no support for the target distro"
-        vm_umount
-        exit 95 # ENOTSUP
-      fi
+  signal_manager 'interrupt_cleanup' || warning 'Was not able to set signal handler'
 
-      modules_install_to "${configurations[mount_point]}" "$flag"
-      ;;
-    2) # LOCAL_TARGET
-      cmd="sudo -E make modules_install"
-      cmd_manager "$flag" "$cmd"
-      ;;
-    3) # REMOTE_TARGET
-      # 1. Preparation steps
-      prepare_host_deploy_dir
+  if [[ "$target" == "$VM_TARGET" ]]; then
+    vm_mount
+    ret="$?"
+    if [[ "$ret" != 0 ]]; then
+      complain "Please shutdown or umount your VM to continue."
+      exit "$ret"
+    fi
+  fi
 
-      remote=$(get_based_on_delimiter "$formatted_remote" ":" 1)
-      port=$(get_based_on_delimiter "$formatted_remote" ":" 2)
-      # User may specify a hostname instead of bare IP
-      remote=$(get_based_on_delimiter "$remote" "@" 2)
+  # NOTE: If we deploy a new kernel image that does not match with the modules,
+  # we can break the boot. For security reason, every time we want to deploy a
+  # new kernel version we also update all modules; maybe one day we can change
+  # it, but for now this looks the safe option.
+  start=$(date +%s)
+  modules_install "" "$target" "$remote"
+  end=$(date +%s)
+  runtime=$((end - start))
 
-      prepare_remote_dir "$remote" "$port" "" "$flag"
+  if [[ "$modules" == 0 ]]; then
+    start=$(date +%s)
+    # Update name: release + alias
+    name=$(make kernelrelease)
 
-      # 2. Send files modules
-      modules_install_to "$KW_CACHE_DIR/$LOCAL_REMOTE_DIR/" "$flag"
+    kernel_install "$reboot" "$name" "" "$target" "$remote"
+    end=$(date +%s)
+    runtime=$((runtime + (end - start)))
+    statistics_manager "deploy" "$runtime"
+  else
+    statistics_manager "Modules_deploy" "$runtime"
+  fi
 
-      release=$(get_kernel_release "$flag")
-      success "Kernel: $release"
-      generate_tarball "$release" "" "$flag"
+  if [[ "$target" == "$VM_TARGET" ]]; then
+    # Umount VM if it remains mounted
+    vm_umount
+  fi
 
-      local tarball_for_deploy_path="$KW_CACHE_DIR/$LOCAL_TO_DEPLOY_DIR/$release.tar"
-      cp_host2remote "$tarball_for_deploy_path" \
-        "$REMOTE_KW_DEPLOY" "$remote" "$port" "" "$flag"
+  if [[ "$target" == "$REMOTE_TARGET" ]]; then
+    say "Cleanup temporary files"
+    cleanup_after_deploy
+  fi
+}
 
-      # 3. Deploy: Execute script
-      local cmd="bash $REMOTE_KW_DEPLOY/deploy.sh --modules $release.tar"
-      cmd_remotely "$cmd" "$flag" "$remote" "$port"
-      ;;
-  esac
+function deploy_help()
```

Comment:
How about putting this `deploy_help` at the end of this file?

Reply:
the whole point of reordering was to keep the functions in order of appearance, I would rather not move `deploy_help`

Reply:
What is the value of seeing such a function in the middle of the code? Technically, this function does not have any value from the development perspective, and if we keep it in the middle, it just looks like a "noise" when I'm looking for something more substantial. That was the exact situation that I have in other functions that have this helper function in the middle of the code.

Reply:
ok, I will move it